### PR TITLE
TokenStream::extend

### PR DESCRIPTION
**Pull request by dtolnay**
Two new insta-stable impls in libproc_macro:

```rust
impl Extend<TokenTree> for TokenStream
impl Extend<TokenStream> for TokenStream
```

`proc_macro::TokenStream` already implements `FromIterator<TokenTree>` and `FromIterator<TokenStream>` so I elected to support the same input types for `Extend`.

**This commit reduces compile time of Serde derives by 60% (takes less than half as long to compile)** as measured by building our test suite:

```console
$ git clone https://github.com/serde-rs/serde
$ cd serde/test_suite
$ cargo check --tests --features proc-macro2/nightly
$ rm -f ../target/debug/deps/libtest_*.rmeta
$ time cargo check --tests --features proc-macro2/nightly
Before: 20.8 seconds
After: 8.6 seconds
```

r? @alexcrichton

### CODE REVIEW

File: src/libsyntax/util/rc_vec.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        // By default, constructing RcVec from Vec gives it just enough capacity
+        // to hold the initial elements. Callers that anticipate needing to
+        // extend the vector may prefer RcVec::new_preserving_capacity.
+        vec.shrink_to_fit();
+
+        RcVec {
+            offset: 0,
+            len: vec.len() as u32,
+            data: Lrc::new(vec),
+        }
```
Comment (alexcrichton) [inline src/libsyntax/util/rc_vec.rs]: Could this delegate to `new_preserving_capacity` after the `shrink_to_fit`?

File: src/libsyntax/util/rc_vec.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    /// underlying vector. Otherwise returns self unmodified.
+    pub fn try_unwrap(self) -> Result<Vec<T>, Self> {
+        match Lrc::try_unwrap(self.data) {
+            // If no other RcVec shares ownership of this data.
+            Ok(mut vec) => {
+                // Drop any elements after our view of the data.
+                vec.truncate(self.offset as usize + self.len as usize);
+                // Drop any elements before our view of the data.
+                if self.offset != 0 {
+                    vec.drain(..self.offset as usize);
```
Comment (alexcrichton) [inline src/libsyntax/util/rc_vec.rs]: Switching these two may be a bit more readable perhaps?

```rust
vec.drain(..self.offset as usize);
vec.truncate(self.len as usize);
```

Additionally, how come the `drain` is conditional?

**Comment by alexcrichton** [comment]
@bors: delegate=dtolnay

Looks great to me, very nice find! r=me with a few nits, and otherwise cc @rust-lang/libs, @nrc, and @petrochenkov on the API addition here.

**Review by petrochenkov — APPROVED** [review]

File: src/libsyntax/util/rc_vec.rs (hunk 1 above)
Comment (dtolnay) [reply src/libsyntax/util/rc_vec.rs]: Done.

File: src/libsyntax/util/rc_vec.rs (hunk 2 above)
Comment (dtolnay) [reply src/libsyntax/util/rc_vec.rs]: I removed the conditional but I kept the order of truncate first then drain. I added a comment to explain -- draining first would involve unnecessarily shifting all the vector elements that are past the end of the current rc's view of the data.

Comment (alexcrichton) [reply src/libsyntax/util/rc_vec.rs]: Aha makes sense 👍

**Comment by BurntPizza** [comment]
@dtolnay I'd like to hear how you found this; just profiling while building a `serde` project?

**Comment by dtolnay** [comment]
@BurntPizza no profiling involved this time, although that would have revealed the problem. I think I've built enough of Serde / syn / quote / proc-macro2 to know where the skeletons are hidden and I finally got a chance to pursue this one.

The original code in [alexcrichton/proc-macro2#114](https://github.com/alexcrichton/proc-macro2/pull/114/files) performs a quadratic amount of memory allocation and copying when called the way `quote!` ends up calling it. I had brought up the need for `Extend` in https://github.com/rust-lang/rust/issues/38356#issuecomment-387935436 during the design phase of Syn 0.14 and the quadratic approach was done as a stopgap until the impl could be written.

**Comment by BurntPizza** [comment]
Ah, the "Because I'm a guru" methodology.
I wish I could apply it more often myself. ;)

**Review by dtolnay — COMMENTED** [review]
Perf looks clean, I see no regression from changing RcSlice to RcVec. [perf comparison](https://perf.rust-lang.org/compare.html?start=50503497492e9bab8bc8c5ad3fe403a3a80276d3&end=b5590423e6ceb048dd7d792382e960d66b7615d2)

**Pull request merged**